### Brief Overview 

- The goal of this assignment is to practice some important data wrangling functionality.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product here is a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.

In [1]:
import pandas as pd
import numpy as np

# 1 Working with the Tax Data


First we load the IRS Statistics of Income (SOI) dataset (tax_data.csv) into a `DataFrame` called `tax_data`. This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [3]:
tax_data = pd.read_csv('~/Downloads/tax_data.csv')

tax_data.head(8)

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 131 columns]

Here we change all the columns names to uppercase.

In [4]:
tax_data.columns = tax_data.columns.str.upper()

print(tax_data.columns.values)

['STATEFIPS' 'STATE' 'ZIPCODE' 'AGI_STUB' 'N1' 'MARS1' 'MARS2' 'MARS4'
 'PREP' 'N2' 'NUMDEP' 'TOTAL_VITA' 'VITA' 'TCE' 'VITA_EIC' 'RAL' 'RAC'
 'ELDERLY' 'A00100' 'N02650' 'A02650' 'N00200' 'A00200' 'N00300' 'A00300'
 'N00600' 'A00600' 'N00650' 'A00650' 'N00700' 'A00700' 'N00900' 'A00900'
 'N01000' 'A01000' 'N01400' 'A01400' 'N01700' 'A01700' 'SCHF' 'N02300'
 'A02300' 'N02500' 'A02500' 'N26270' 'A26270' 'N02900' 'A02900' 'N03220'
 'A03220' 'N03300' 'A03300' 'N03270' 'A03270' 'N03150' 'A03150' 'N03210'
 'A03210' 'N03230' 'A03230' 'N03240' 'A03240' 'N04470' 'A04470' 'A00101'
 'N18425' 'A18425' 'N18450' 'A18450' 'N18500' 'A18500' 'N18300' 'A18300'
 'N19300' 'A19300' 'N19700' 'A19700' 'N04800' 'A04800' 'N05800' 'A05800'
 'N09600' 'A09600' 'N05780' 'A05780' 'N07100' 'A07100' 'N07300' 'A07300'
 'N07180' 'A07180' 'N07230' 'A07230' 'N07240' 'A07240' 'N07220' 'A07220'
 'N07260' 'A07260' 'N09400' 'A09400' 'N85770' 'A85770' 'N85775' 'A85775'
 'N09750' 'A09750' 'N10600' 'A10600' 'N59660' 'A59660' '

Let's see how many observations are in this data set.

In [5]:
print(f'The total number of obervations is {tax_data.shape[0]}')

The total number of obervations is 166698


Here we print the count of unique zip codes in each state in descending order.

In [6]:
tax_data_nzipcodes_per_state = (tax_data[['STATE','ZIPCODE']].groupby(['STATE'])
                                                             .nunique()
                                                             .sort_values(by=['ZIPCODE'], ascending=False)
                                                             .drop(['STATE'], axis=1))
tax_data_nzipcodes_per_state

ZIPCODE
STATE         
TX        1619
NY        1541
CA        1486
PA        1368
IL        1231
OH         998
FL         920
MI         892
MO         890
IA         827
VA         793
MN         792
NC         725
WI         714
IN         674
GA         667
KY         656
KS         598
TN         590
AL         577
NJ         547
OK         547
WV         512
WA         497
AR         490
NE         486
MA         482
LA         453
MD         405
CO         395
SC         376
MS         370
ME         370
OR         356
AZ         295
ND         287
SD         287
CT         264
VT         238
NH         232
MT         225
ID         213
NM         209
UT         186
NV         128
WY         109
RI          71
HI          60
DE          57
AK          56
DC          24

As shown below, HI is at the 48th position in the list.

In [7]:
position = tax_data_nzipcodes_per_state.index.get_loc('HI') + 1

print(f'HI is at the {position}th position in the list, and TX is at the first.')

HI is at the 48th position in the list, and TX is at the first.


### Identifying and Removing Ambiguous Zip Codes

First we count the number of entries where ZIPCODE is 0, and assign the results to `nb_invalid_zip`

In [8]:
nb_invalid_zip = (tax_data.ZIPCODE.eq(0)).sum().item()

Next we remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to `tax_data_valid_zip`

In [10]:
tax_data_valid_zip = tax_data[tax_data.ZIPCODE.ne(0)]

The assertion below is testing that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
It will fail (and print an error message) if the results do not match.

In [11]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

Here we find how many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`, and assign the count of `NaN` to `nb_missing_values`

In [12]:
nb_missing_values = (tax_data_valid_zip.isnull()
                                       .sum(axis=1)
                                       .ge(1)
                                       .sum())

print(f'{nb_missing_values} lines contain at least one missing value.')

199 lines contain at least one missing value.


Next we create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values, and call it `tax_data_valid_zip_cleaned`

In [13]:
tax_data_valid_zip_cleaned = tax_data_valid_zip[tax_data_valid_zip.isnull().sum(axis=1).eq(0)]

The assertion below is testing that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`

In [16]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code

* The function `compute_percentile_zipcode` below computes the percentile income per zip code

* By default percentile=0.5,  i.e., the function computes the median

In [15]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

Here we compute the 65th income percentile ( percentile=0.65) for each zipcode in `tax_data_valid_zip_cleaned`, and we sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`

In [16]:
tax_data_valid_zip_cleaned.loc[:,'ZIPCODE'] = tax_data_valid_zip_cleaned.loc[:,'ZIPCODE'].astype('int')

zip_rev_all = (tax_data_valid_zip_cleaned.groupby(['ZIPCODE'])
                                         .apply(compute_percentile_zip, 0.65)
                                         .sort_values(ascending=False))
zip_rev_all

ZIPCODE
33109    3954.114286
33480    3413.301538
94301    3109.443711
94027    3091.537013
10577    2414.855556
            ...     
53706       7.292857
55455       7.210000
48109       6.981250
84112       6.597297
59053       5.421429
Length: 27682, dtype: float64

Here we find the three zip codes with the most significant 65th percentile value for income?

In [17]:
print(f'{zip_rev_all.index[0]}, {zip_rev_all.index[1]}, and {zip_rev_all.index[2]} are the three zip codes with the most significant 65th percentile value for income.')

33109, 33480, and 94301 are the three zip codes with the most significant 65th percentile value for income.


# 2 Working with the Medicaid Data

### Loading and exploring the data 

First we load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`.

In [18]:
medicaid_data = pd.read_csv('~/Downloads/medicaid_data.csv')

medicaid_data.head()

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)

Next we convert all the column names to uppercase. Then we check that the operation worked along with checking the number of rows.

In [20]:
medicaid_data.columns = medicaid_data.columns.str.upper()

print(f'Number of Rows: {medicaid_data.shape[0]}\n')
print(f'Column Headers: {medicaid_data.columns.values}')

Number of Rows: 1695546

Column Headers: ['UTILIZATION TYPE' 'STATE' 'NDC' 'PRODUCT NAME' 'UNITS REIMBURSED'
 'NUMBER OF PRESCRIPTIONS' 'TOTAL AMOUNT REIMBURSED'
 'MEDICAID AMOUNT REIMBURSED' 'NON MEDICAID AMOUNT REIMBURSED' 'LOCATION']


Are there any states that have more than one value for `Location`?

In [23]:
# Keep only unique rows of the dataframe medicaid_data[['STATE', 'LOCATION']], and sort by STATE
# Then convert to a Series using just the STATE column
# If a state has more than one value for LOCATION, it will be listed more than once on the Series
medicaid_data_states_drop_duplicate_locations = (medicaid_data[['STATE', 'LOCATION']]
                                                 .sort_values(by=['STATE', 'LOCATION'])
                                                 .drop_duplicates()
                                                 .STATE)

# A sorted Series of each state that appears in the data frame
# Each state name appears only once
medicaid_data_states = medicaid_data.STATE.sort_values().drop_duplicates()

# If there exists a state that has more than one value for LOCATION, total will be greater than 0
# Otherwise, total will equal 0
total = (medicaid_data_states_drop_duplicate_locations.reset_index(drop=True)
         .ne(medicaid_data_states.reset_index(drop=True))
         .sum())


print(f'Since total = {total}, then we know that there are no states that have more than one value for LOCATION.')

Since total = 0, then we know that there are no states that have more than one value for LOCATION.


To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).
  
        https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

The number of enrollees per state was obtained here:

        https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html
    
    
Here we use a parsed/processed version called `medicaid_enrollment.tsv` 

In [22]:
medicaid_enrollment = pd.read_csv('~/Downloads/medicaid_enrollment.tsv', sep='\t')

medicaid_enrollment.head()

STATE Total Medicaid Enrollees
0         Alabama                1,050,989
1          Alaska                  164,783
2  American Samoa                      NaN
3         Arizona                1,740,520
4        Arkansas                  762,166

We convert the column names to uppercase.

In [23]:
medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()

Some states/territories have missing values. So let's remove the missing values and save the resulting `DataFrame` to `medicaid_enrollment_cleaned`

Some of the missing values are coded as 'n/a'. So in order for our operation to work, we have to convert this value to a value that is recognized by the `dropna()` function.

In [24]:
# Replace white spaces and 'n/a'
medicaid_enrollment[['TOTAL MEDICAID ENROLLEES']] = ((medicaid_enrollment[['TOTAL MEDICAID ENROLLEES']])
                                                      .replace(to_replace=[r'\s', 'n/a'], 
                                                               value=['', np.NaN], 
                                                               regex=True))

# Delete rows with missing values
medicaid_enrollment_cleaned = medicaid_enrollment.dropna().reset_index(drop=True)

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization.

We then inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and  to sure that the data type is `int`

In [25]:
# Replace commas
medicaid_enrollment_cleaned[['TOTAL MEDICAID ENROLLEES']] = ((medicaid_enrollment_cleaned[['TOTAL MEDICAID ENROLLEES']])
                                                            .replace(to_replace=',', value='', regex=True))

# Convert to int
medicaid_enrollment_cleaned[['TOTAL MEDICAID ENROLLEES']] = (medicaid_enrollment_cleaned[['TOTAL MEDICAID ENROLLEES']]
                                                             .astype('int64'))

print(f"The data type is {medicaid_enrollment_cleaned[['TOTAL MEDICAID ENROLLEES']].dtypes[0]}")

The data type is int64


### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

We can use the shared State information across both tables to associate both tables. However,  `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name. Therefore, we need to convert (or append) the two-letter state abbreviations to `medicaid_enrollment_cleaned`

In [26]:
import requests
from io import BytesIO

headers = { 'Accept' : '*/*',
            'User-Agent' : 'Mozilla/5.0',
            'Refers' : 'http://www.nseindia.com',
            'Connection' : 'keep-alive'
          }

contents = requests.get("https://www.50states.com/abbreviations.htm", headers=headers)


# reads all the tables at the given url
tables = pd.read_html(BytesIO(contents.content))



# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]

# Add Puerto Rico
Codes_abbreviations = Codes_abbreviations.append({'US State:': 'Puerto Rico', 'Abbreviation:': 'PR'}, ignore_index=True)
Codes_abbreviations.head()

US State: Abbreviation:
0     Alabama            AL
1      Alaska            AK
2     Arizona            AZ
3    Arkansas            AR
4  California            CA

And of course, we change the `DataFrame`'s headers to uppercase.

In [27]:
Codes_abbreviations.columns = ['US STATE', 'ABBREVIATION']

Next we combine `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations`, and we save the results to `medicaid_enrollment_cleaned_with_zip` 

In [28]:
# Boolean: Is the abbreviation in medicaid_enrollment?
Codes_abbreviations_isin = Codes_abbreviations['US STATE'].isin(medicaid_enrollment_cleaned.STATE)

# Drop rows that have abbreviations not in medicaid_enrollment
Codes_abbreviations_cleaned = (Codes_abbreviations[Codes_abbreviations_isin].sort_values(by='US STATE')
                                                                            .reset_index(drop=True))
# Join
medicaid_enrollment_cleaned_with_zip = (medicaid_enrollment_cleaned.join(Codes_abbreviations_cleaned)
                                                                   .drop('US STATE', axis=1))

medicaid_enrollment_cleaned_with_zip.head()

STATE  TOTAL MEDICAID ENROLLEES ABBREVIATION
0     Alabama                   1050989           AL
1      Alaska                    164783           AK
2     Arizona                   1740520           AZ
3    Arkansas                    762166           AR
4  California                  13096861           CA

We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`, so we remove it.

In [29]:
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop('STATE', axis=1)
medicaid_enrollment_cleaned_with_zip.head()

TOTAL MEDICAID ENROLLEES ABBREVIATION
0                   1050989           AL
1                    164783           AK
2                   1740520           AZ
3                    762166           AR
4                  13096861           CA

Next we use `medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`, and we save the resulting DataFrame into `medicaid_data_w_enrollments`

In [30]:
# Boolean: Is state abbreviation from medicaid_..._zip in medicaid_data?
medicaid_zip_isin = medicaid_enrollment_cleaned_with_zip.ABBREVIATION.isin(medicaid_data.STATE)

# Drop rows that have abbreviations not in medicaid_data
medicaid_enrollment_cleaned_with_zip_culled = medicaid_enrollment_cleaned_with_zip[medicaid_zip_isin]

# Merge
medicaid_data_w_enrollments = (medicaid_data.merge(medicaid_enrollment_cleaned_with_zip_culled, 
                                                  left_on='STATE', 
                                                  right_on='ABBREVIATION', 
                                                  validate='m:1')
                                            .drop('ABBREVIATION', axis=1))

medicaid_data_w_enrollments.head()

UTILIZATION TYPE STATE          NDC PRODUCT NAME  UNITS REIMBURSED  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    PA    456060010   TEFLARO 60             224.0   
2             MCOU    PA    115692201   DIVALPROEX            1626.0   
3             FFSU    PA     54455025    METHOTREX             450.0   
4             MCOU    PA  50458014090     INVOKANA            3850.0   

   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
0                     19.0                   234.98   
1                     36.0                 35752.31   
2                     20.0                  2318.55   
3                     22.0                   565.29   
4                    111.0                 52655.13   

   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
0                      234.98                            0.00   
1                    35752.31                            0.00   
2                     2287.80                           30.75   
3                      516.23                           49.06   
4                    50641.22                         2013.91   

             LOCATION  TOTAL MEDICAID ENROLLEES  
0  (40.5773, -77.264)                   2569232  
1  (40.5773, -77.264)                   2569232  
2  (40.5773, -77.264)                   2569232  
3  (40.5773, -77.264)                   2569232  
4  (40.5773, -77.264)                   2569232

Here we remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [31]:
state_values = medicaid_data_w_enrollments.STATE
product_values = medicaid_data_w_enrollments['PRODUCT NAME']

# Compare the list of STATE values to ABBREVIATION coloumn in Codes_abbrevations_cleaned
# Sum the True values. This sum will equal the size of the list of STATE values if there are no missing values
# We do it this way because missing state values are coded as XX. I don't think .isnull() will pick that up.
no_missing_state_values = (state_values.isin(Codes_abbreviations_cleaned.ABBREVIATION).sum() == (state_values.size))

# Check for missing PRODUCT NAME values
no_missing_product_values = product_values.isnull().sum() == 0

print(f'There are no missing STATE values: {no_missing_state_values}')
print(f'There are no missing PRODUCT NAME values: {no_missing_product_values}')

# Find the missing values in the PRODUCT NAMES column and drop them
medicaid_data_w_enrollments = medicaid_data_w_enrollments[product_values.isnull().eq(False)]
no_missing_product_values = medicaid_data_w_enrollments['PRODUCT NAME'].isnull().sum() == 0

print(f'\nAfter dropping the missing PRODUCT NAME values there are no missing PRODUCT NAME values: {no_missing_product_values}')

There are no missing STATE values: True
There are no missing PRODUCT NAME values: False

After dropping the missing PRODUCT NAME values there are no missing PRODUCT NAME values: True


We now use ["STATE", "NDC"] as hierarchical index for `medicaid_data_w_enrollments`, and we call the new DataFrame `medicaid_data_w_enrollments_hierarch`

In [32]:
# Sort medicaid by STATE and NDC
medicaid_data_w_enrollments.sort_values(by=['STATE', 'NDC'], inplace=True)

# Make medicaid hierarch dataframe with STATE and NDC as indices 
medicaid_data_w_enrollments_hierarch = medicaid_data_w_enrollments
medicaid_data_w_enrollments_hierarch.index = [medicaid_data_w_enrollments.STATE, medicaid_data_w_enrollments.NDC]

print(f'The new index now has the two levels: {medicaid_data_w_enrollments_hierarch.index.names}')

# Drop STATE and NDC columns
medicaid_data_w_enrollments_hierarch.drop(['STATE', 'NDC'], axis=1, inplace=True)
medicaid_data_w_enrollments_hierarch

The new index now has the two levels: ['STATE', 'NDC']


UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
AK    2143380                 FFSU    TRULICITY              57.0   
      2143380                 FFSU    TRULICITY              58.0   
      2143380                 FFSU    TRULICITY              96.0   
      2143480                 FFSU    TRULICITY              46.0   
      2143480                 FFSU    TRULICITY              56.0   
...                            ...          ...               ...   
WY    76385010401             FFSU   Benztropin            1020.0   
      76385010401             FFSU   Benztropin             780.0   
      76385010501             FFSU   Benztropin            1065.0   
      76439030910             FFSU   HYOSCYAMIN            1214.0   
      76439030910             FFSU   HYOSCYAMIN            1184.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
AK    2143380                         28.0                 17493.58   
      2143380                         28.0                 18439.17   
      2143380                         48.0                 31466.12   
      2143480                         23.0                 14096.17   
      2143480                         28.0                 17996.75   
...                                    ...                      ...   
WY    76385010401                     21.0                   299.72   
      76385010401                     12.0                   148.80   
      76385010501                     22.0                   334.64   
      76439030910                     20.0                   467.07   
      76439030910                     22.0                   485.54   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
AK    2143380                        17493.58                             0.0   
      2143380                        18439.17                             0.0   
      2143380                        31466.12                             0.0   
      2143480                        14096.17                             0.0   
      2143480                        17996.75                             0.0   
...                                       ...                             ...   
WY    76385010401                      299.72                             0.0   
      76385010401                      148.80                             0.0   
      76385010501                      334.64                             0.0   
      76439030910                      467.07                             0.0   
      76439030910                      485.54                             0.0   

                               LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                          
AK    2143380       (61.385, -152.2683)                    164783  
      2143380       (61.385, -152.2683)                    164783  
      2143380       (61.385, -152.2683)                    164783  
      2143480       (61.385, -152.2683)                    164783  
      2143480       (61.385, -152.2683)                    164783  
...                                 ...                       ...  
WY    76385010401  (42.7475, -107.2085)                     66532  
      76385010401  (42.7475, -107.2085)                     66532  
      76385010501  (42.7475, -107.2085)                     66532  
      76439030910  (42.7475, -107.2085)                     66532  
      76439030910  (42.7475, -107.2085)                     66532  

[1542292 rows x 9 columns]

Let's print all the lines containing NDC 61958180101 in "PA"

In [33]:
medicaid_data_w_enrollments_hierarch.loc[('PA', 61958180101)]

UTILIZATION TYPE PRODUCT NAME  UNITS REIMBURSED  \
STATE NDC                                                           
PA    61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            2604.0   
      61958180101             FFSU   Harvoni  (            1008.0   
      61958180101             MCOU   Harvoni  (            1932.0   
      61958180101             FFSU   Harvoni  (             924.0   
      61958180101             MCOU   Harvoni  (            3220.0   

                   NUMBER OF PRESCRIPTIONS  TOTAL AMOUNT REIMBURSED  \
STATE NDC                                                             
PA    61958180101                     33.0               1017644.54   
      61958180101                     93.0               2952891.16   
      61958180101                     36.0               1107725.22   
      61958180101                     69.0               2178602.30   
      61958180101                     33.0               1015383.60   
      61958180101                    115.0               3599192.35   

                   MEDICAID AMOUNT REIMBURSED  NON MEDICAID AMOUNT REIMBURSED  \
STATE NDC                                                                       
PA    61958180101                   985269.84                        32374.70   
      61958180101                  2860100.54                        92790.62   
      61958180101                  1107725.22                            0.00   
      61958180101                  2118612.30                        59990.00   
      61958180101                  1015383.60                            0.00   
      61958180101                  3482188.68                       117003.67   

                             LOCATION  TOTAL MEDICAID ENROLLEES  
STATE NDC                                                        
PA    61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232  
      61958180101  (40.5773, -77.264)                   2569232

#### Computing the normalized `UNITS REIMBURSED` per drug

First, lets compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.

- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612

```python
total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 
```

- And the number of Medicaid Enrollees in "PA" is 2,569,232


```python
total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```
- Therefore, the UNITS REIMBURSED per enrollee  for "HARVONI" is  0.00413041718303


```python
print(total_amount_reimbursed/total_enrollees_AK)
```

Rather than work directly with the ratios, we are going to compute and work with their logarithm (np.log2) instead, and save the result in a `Series` called `medicaid_reimbursement_per_enrollee`

In [34]:
# Get subset of medicaid_data_..._hierarch with reset index
medicaid_hierarch_sub = (medicaid_data_w_enrollments_hierarch[['UNITS REIMBURSED', 'TOTAL MEDICAID ENROLLEES']]
                         .reset_index())

# Sum the Units Reiembursed for each NDC
units_reimbursed_sum = (medicaid_hierarch_sub[['STATE', 'NDC', 'UNITS REIMBURSED']]
                        .groupby(['STATE', 'NDC'])
                        .sum())

# Total Medicaid Enrollees
total_enrollees = (medicaid_data_w_enrollments_hierarch['TOTAL MEDICAID ENROLLEES']
                   .reset_index()
                   .drop_duplicates()
                   .reset_index(drop=True))

# Index Total Mediciaid Enrollees with [STATE, NDC]
total_enrollees.index = units_reimbursed_sum.index

# Drop STATE  and NDC columns
total_enrollees = total_enrollees['TOTAL MEDICAID ENROLLEES']

# Join
units_reiembursed_total_enrollees_join = units_reimbursed_sum.join(total_enrollees)

# Compute log-normalized ratio of UNITS REIMBURSED
medicaid_reimbursement_per_enrollee = (units_reiembursed_total_enrollees_join['UNITS REIMBURSED']
                                       .div(units_reiembursed_total_enrollees_join['TOTAL MEDICAID ENROLLEES']
                                            .values, 
                                            axis=0)
                                       .transform(np.log2))

medicaid_reimbursement_per_enrollee

STATE  NDC        
AK     2143380        -9.609109
       2143480       -10.008280
       2322730        -6.109830
       2322830        -4.444321
       2322930        -3.855995
                        ...    
WY     76329336901    -9.773833
       76385010301    -6.146779
       76385010401    -4.707744
       76385010501    -5.965123
       76439030910    -4.794145
Length: 473016, dtype: float64

#### Exploring the data (very briefly) 
- What is the drug with the highest log-normalized ratio in Hawaii?


In [35]:
# The index of the highest log-normalized ratio in Hawaii
max_ratio_index = medicaid_reimbursement_per_enrollee.loc['HI'].idxmax()

# The drug name
drug = medicaid_data_w_enrollments_hierarch.loc[('HI', max_ratio_index), 'PRODUCT NAME'][0]

print(f'{drug} is the drug with the highest log-normalized ratio in Hawaii.')

GAVILYTE-G is the drug with the highest log-normalized ratio in Hawaii.


**Gavilyte-G** is a medical grade laxative. Its primarily use is as a colon cleanse before getting a colonoscopy, so that the doctor can see the details of the colon more clearly.

In [39]:
# Rows of medicaid_data_w_enrollments_hierarch with the desired product name
gavilyte = (medicaid_data_w_enrollments_hierarch[medicaid_data_w_enrollments_hierarch['PRODUCT NAME']
                                                 .eq('GAVILYTE-G')][['PRODUCT NAME', 'UNITS REIMBURSED']])

# Sumed Units Reimbured for HI and other indicated states, with the desired product name
(gavilyte[gavilyte.index
                  .droplevel('NDC')
                  .isin(['AZ', 'FL', 'HI', 'MA', 'TX', 'VA'])]
 .sum(level='STATE'))

UNITS REIMBURSED
STATE                  
AZ           20928400.0
FL            1848000.0
HI          496952018.0
MA               5433.0
TX               3869.0
VA            3428000.0

Considering that Gavilyte-G's main use is to prepare a patient for a colonoscopy by flushing out their bowls, the high ratio must mean that a higher percentage of Hawaii Medicaid enrollees are getting colonoscopies. One of the main reasons to get a colonoscopy is to check for colon cancer. I'm not too versed on medical matters, but in my opinion, the high number could be a good thing or a bad thing.
<br>
<br>
It's a good thing if the colonoscopies are being ordered as a precaution. The precaution being that the doctor can look for the early signs of colon cancer and begin treatment early on before the cancer becomes severe.
<br>
<br>
It's a bad thing if the main reason for all the colonoscopies is because Hawaii is trying to get a handle on some sort of colon cancer epidemic.

Next we find and list all unique `NDC`s for which the difference between the largest and second largest log-normalized ratio by state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio  of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ration of  `-12.623428`
  


In [36]:
# Get a series of the 2 largest ratios, grouped by NDC
series_2largest = medicaid_reimbursement_per_enrollee.groupby(level=1).nlargest(2)
series_2largest.index = series_2largest.index.droplevel(2)

# Loop on unique NDCs to make a visually appealing DataFrame of ratios with difference >= 10
columns=['NDC', 'NAME', 'MAX1', 'STATE1', 'MAX2', 'STATE2']
ndc_df = pd.DataFrame([], columns=columns)
ndc_index = series_2largest.index.droplevel(1).drop_duplicates()
for ndc in ndc_index:
    
    if (series_2largest.loc[(ndc,)].size == 2):
        max1 = series_2largest.loc[(ndc,)][0]
        max2 = series_2largest.loc[(ndc,)][1]
        diff = max1 - max2
        abs_val = np.abs(diff)
        
        if (abs_val >= 10):
            state1 = series_2largest.loc[(ndc,)].index[0]
            state2 = series_2largest.loc[(ndc,)].index[1]
            name = medicaid_data_w_enrollments_hierarch['PRODUCT NAME'].loc[(state1, ndc)].unique()[0]
            ndc_df = ndc_df.append(pd.DataFrame([[ndc, name, max1, state1, max2, state2]], columns=columns), 
                                   ignore_index=True)
            
ndc_df.sort_values(by='MAX1', ascending=False).reset_index(drop=True)

NDC        NAME       MAX1 STATE1       MAX2 STATE2
0   68180062210  CEFTRIAXON   9.668592     VA  -6.845952     NV
1     409767009  NORMOSOL-R   0.366216     NV -15.544686     CA
2     574082710  TESTOSTERO  -0.457059     OK -12.006665     MT
3     591289749  AZACITIDIN  -1.025642     OK -12.623428     GA
4   60793060002  BICILLIN C  -3.050964     OK -15.074396     MN
5     409796709  NORMOSOL-R  -3.655656     NE -15.238426     IA
6     338007304  2.5% DEXTR  -3.732656     WI -17.942278     CA
7     406036501  HYDROCODON  -4.083674     MT -15.041465     MO
8   39822420006  ROCURONIUM  -4.732155     AK -17.216453     CA
9   68001028529  LEUCOVORIN  -5.497791     OK -15.894488     NJ
10    264761210  DEXTROSE 5  -6.738824     TX -19.835363     CA
11  63323017530  Oxaliplati  -7.524275     SD -19.183286     CA
12    409710002  DEXTROSE 5  -7.695279     IN -19.735827     CA
13    641047721  PHENOBARBI  -7.739808     AK -18.327441     CA
14      9023303  BACITRACIN  -8.807845     MO -20.441084     CA
15  50458016601    LEVAQUIN  -9.646134     LA -20.292431     KY
16    407141233   OMNIPAQUE -11.535937     MO -23.977254     MI

- The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states.
   - Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   - Second highest state has a log value of -12.623428 (0.000158478197834722)
- Oklahoma is not a high-incidence state for cancer
- Could the following explain what is happening in Oklahoma?

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10